modules/cluster-management/src/main/java/org/apache/ignite/internal/cluster/management/ClusterManagementGroupManager.java

```diff
@@ -133,90 +138,377 @@ public void initCluster(Collection<String> metaStorageNodeNames, Collection<Stri
+
+        return startCmgRaftService(cmgNodes)
+                .thenCompose(service -> {
+                    log.info("Local CMG state recovered");
+
+                    return service.isCurrentNodeLeader()
+                            .thenCompose(isLeader -> {
+                                if (isLeader) {
+                                    return service.readClusterState()
+                                            // Raft state might not have been initialized in case of leader failure during cluster init
+                                            .thenCompose(state -> state == null ? completedFuture(null) : onLeaderElected(service, state));
+                                } else {
+                                    return completedFuture(null);
+                                }
+                            })
+                            .thenApply(v -> service);
+                });
+    }
+
+    /**
+     * Handles the Init command.
+     *
+     * <p>This method needs to take the following possibilities into account, depending on the local state and the Raft state:
+     * <ol>
+     *     <li>No local state found - this means that the current node has never been initialized before.</li>
+     *     <li>Local state found, but no CMG state present in the Raft storage - this means that the node has failed somewhere during
+     *     the init process. In this case we need to check the consistency of the local state and the received message and complete
+     *     the init process.</li>
+     *     <li>Local state found and CMG state is present in the Raft storage - this means that the node has been initialized successfully
+     *     and a user may be retrying the init in case the successful response was lost. To make the init message idempotent
+     *     we simply check that the Raft state and the received message are the same.</li>
+     * </ol>
+     */
+    private void handleInit(CmgInitMessage msg, NetworkAddress addr, long correlationId) {
+        var newState = new ClusterState(msg.cmgNodes(), msg.metaStorageNodes());
+
+        CompletableFuture<CmgRaftService> raftService = this.raftService;
+
+        // This future is needed to add a completion listener at the end of the method
+        CompletableFuture<?> resultHook;
+
+        if (raftService == null) {
+            // Raft service has not been started
+            log.info("Init command received, starting the CMG: " + newState);
+
+            raftService = initCmgRaftService(newState);
+
+            this.raftService = raftService;
+
+            resultHook = raftService;
+        } else {
+            // Raft service has been started, which means that this node has already received an init command at least once, but
+            // we still need to check that the initialization has completed successfully.
+            log.info("Init command received, but the CMG has already been started");
+
+            resultHook = raftService.thenCompose(service ->
```

Comment:
No, because in the first `if` branch the Raft service is not started yet and we start it there. In this branch the Raft service has already been started (and this future only represents the state of the Raft service), so we only add some async hooks that should send a response to the user upon completion.